### ISSUE TITLE
Top Level Await

### ISSUE THREAD

**Issue by rwaldron** [tc39/test262#2218]
ref: https://tc39.es/proposal-top-level-await/

@leobalter please fill in the following work lists and commence work on new test material: 

- [ ]  [AsyncFunctionStart ( promiseCapability, asyncFunctionBody )](https://tc39.es/proposal-top-level-await/#sec-async-functions-abstract-operations-async-function-start)
  - [ ] ...
- [ ]  [AsyncBlockStart ( promiseCapability, asyncBody, asyncContext )](https://tc39.es/proposal-top-level-await/#sec-asyncblockstart)
  - [ ] ...
- [ ]  [Abstract Module Records](https://tc39.es/proposal-top-level-await/#sec-abstract-module-records)
  - [ ] ...
- [ ]  [Cyclic Module Records](https://tc39.es/proposal-top-level-await/#sec-cyclic-module-records)
  - [ ] ...

- [ ] Table 3: Additional Fields of Cyclic Module Records

  - [ ] [[Async]] 
    - [ ] ...
  - [ ] [[AsyncEvaluating]] true or false Whether this module is currently awaiting async fulfillment.
    - [ ] ...
  - [ ] [[TopLevelCapability]]
    - [ ] ...
  - [ ] [[AsyncParentModules]]
    - [ ] ...
  - [ ] [[PendingAsyncDependencies]]
    - [ ] ...

- [ ] Table 4: Additional Abstract Methods of Cyclic Module Records
  - [ ] ExecuteModule( [ promiseCapability ] )
    - [ ] ... 
- [ ] [Link ( ) Concrete Method](https://tc39.es/proposal-top-level-await/#sec-moduledeclarationlinking)
  - [ ] ...
- [ ] InnerModuleLinking ( module, stack, index )
  - [ ] ...
- [ ] Evaluate ( ) Concrete Method
  - [ ] ...
- [ ] InnerModuleEvaluation( module, stack, index )
  - [ ] ...
- [ ] ExecuteAsyncModule ( module )
  - [ ] ...
- [ ] GetCycleRoot ( module )
  - [ ] ...
- [ ] AsyncModuleExecutionFulfilled ( module )
  - [ ] ...
- [ ] AsyncModuleExecutionRejected ( module, error )
  - [ ] ...
- [ ] Source Text Module Records
  - [ ] ...
- [ ] ParseModule ( sourceText, realm, hostDefined )
  - [ ] ...
- [ ] ExecuteModule ( [ capability ] )
  - [ ] ...

- [ ] Runtime Semantics: TopLevelModuleEvaluationJob ( sourceText, hostDefined )
  - [ ] ...

- [ ] Modules

```
StatementListItem[~Yield, +Await, ~Return]
```

- [ ] Async Function Definitions (Syntax)
  - [ ] ...
- [ ] Imports
  - [ ] ...
- [ ] Exports
  - [ ] ...

Note: There may be top level sections missings, so please keep an eye out for that.

**Comment by mb21** [comment]
Hello! This issue mentions "Cyclic Module Records", but being unfamiliar with this testsuite, I'm unsure whether in the end any tests regarding that were added? The reason I'm asking is that it would be nice to reference a test that would be failing in Safari for https://github.com/web-platform-tests/interop/issues/1105, thanks!

**Comment by ptomato** [comment]
@mb21 Sorry, none of the maintainers from 2019 are still involved with the project. At this point your best bet is to take a look at which toplevel await tests fail in JavaScriptCore and then see if one covers the case you are interested in.

You can use [ESVU](https://github.com/devsnek/esvu) to install a recent JSC standalone interpreter. The incantation that will get you the JSC results for all toplevel await tests is
```
npx test262-harness --hostType=jsc --hostPath=$HOME/.esvu/bin/jsc -f top-level-await "test/**"
```
Also check the tests in #4465, #4500, #4501, #4502 which may be related to what you are looking for.

**Comment by mb21** [comment]
@ptomato thanks a lot!

For that incantation to work, I first need to `npx esvu install jsc `, correct? Unfortunately, I get:

> Fatal error installing JavaScriptCore Error: Unsupported macOS release: sequoia

Since there doesn't seem to be a way to install older versions of an engine, seems I first have to update macOS? Well yes, let's see when I get around to that. 😅 

I don't suppose this testsuite is run and listed somewhere, similar to https://wpt.fyi/results/ ?

**Comment by ptomato** [comment]
@mb21 Yes, https://test262.fyi/# - but you won't be able to select for only toplevel await tests.
